prowgen: (further) refactor code for reusability

petr-muller opened the pull request:
The motivation for this refactor is allowing external code
(payload-testing controller, specifically) to reuse Prowgen code without
needing to duplicate parts of it.

The refactor is based on extracting the bits that handle generating the
`JobBase` part of the ProwJob, and making two constructors for new
builder: one that covers the part of the `JobBase` that is generated
from ci-operator config as a whole, and one that covers the parts that
are generated for individual test items in that config. This allows to
simplify the signatures of the `Generate*ForTest` methods, because they
now only handle the Presubmit/Postsubmit/Periodic generation, without
the `JobBase` part they have in common.

The refactor could go even further, there are still duplicities to
eliminate, but I would like to stop here, because this PR carries the
changes necessary for integrating
https://github.com/openshift/ci-tools/pull/2462 work into the
controller. I believe that even in this partial way it improves the
structure of Prowgen code by simplifying some of the huge methods that
were there.

/cc @openshift/test-platform

petr-muller commented on pkg/prowgen/jobbase.go:
```diff
@@ -0,0 +1,185 @@
+package prowgen
```
Most functional code in this file was moved here from `prowgen.go`, especially from the `GenerateJobs` method. It is just structured differently, allowing just parts of the generation code to be used only when needed.

petr-muller commented on pkg/prowgen/jobbase_test.go:
```diff
@@ -0,0 +1,418 @@
+package prowgen
```
Huge file but really just test scenarios - the actual test code is usually super simple.

petr-muller commented on pkg/prowgen/podspec.go:
```diff
@@ -668,3 +668,22 @@ func (c *ciOperatorPodSpecGenerator) MustBuild() *corev1.PodSpec {
+func newFakePodSpecBuilder() CiOperatorPodSpecGenerator {
+	f := fakePodSpecBuilder(0)
+	return &f
+}
```
This is used in testing the JobBaseBuilder which accepts a PodSpecBuilder as injected dependency which allows to generate empty specs in tested JobBases, making fixtures shorter and containing less boilerplate.

petr-muller commented on pkg/prowgen/prowgen.go:
```diff
@@ -46,82 +44,9 @@ func GenerateJobs(configSpec *cioperatorapi.ReleaseBuildConfiguration, info *Pro
-				Template("cluster-launch-installer-custom-test-image", element.Commands, fromImage, element.As, element.OpenshiftInstallerCustomTestImageClusterTestConfiguration.ClusterProfile),
-				LeaseClient(),
-			)
-		}
```
Moved to jobbase builder

petr-muller commented on pkg/prowgen/prowgen.go:
```diff
@@ -132,28 +57,28 @@ func GenerateJobs(configSpec *cioperatorapi.ReleaseBuildConfiguration, info *Pro
 			postsubmits[orgrepo] = append(postsubmits[orgrepo], *postsubmit)
 		} else {
-			presubmit := *generatePresubmitForTest(element.As, info, g.MustBuild(), configSpec.CanonicalGoRepository, jobRelease, skipCloning, element.RunIfChanged, element.SkipIfOnlyChanged, element.Optional, element.Timeout, noBuilds)
+			presubmit := generatePresubmitForTest(g, element.As, info, element.RunIfChanged, element.SkipIfOnlyChanged, element.Optional)
```
All these methods now have simpler signatures because they no longer build `JobBase`s

petr-muller commented on pkg/prowgen/prowgen.go:
```diff
@@ -213,20 +147,6 @@ func GenerateJobs(configSpec *cioperatorapi.ReleaseBuildConfiguration, info *Pro
-	}
-	return false
-}
-
```
Moved to jobbase builder

petr-muller commented on pkg/prowgen/prowgen.go:
```diff
@@ -289,50 +210,3 @@ func generatePeriodicForTest(name string, info *ProwgenInfo, podSpec *corev1.Pod
-		base.Hidden = true
-	}
-	return base
-}
```
Moved to jobbase builder

petr-muller commented on pkg/prowgen/testdata/zz_fixture_TestGenerateJobBase_config_variant.yaml:
```diff
@@ -5,7 +5,3 @@ decoration_config:
-spec:
-  containers:
-  - name: test
-    resources: {}
```
Specs are irrelevant when testing `JobBases` so they don't need to be in the fixtures

petr-muller reviewed (commented):
Left some review hints for reviewers, inline

petr-muller commented on pkg/prowgen/prowgen.go:
```diff
@@ -23,6 +20,7 @@ const (
 	Generator      jc.Generator = "prowgen"
 )
 
+//goland:noinspection GoNameStartsWithPackageName
```
I can remove it, I just hate how Goland flags this for me as a warning :P

droslean replied:
I hate goland.... I use vscode

petr-muller commented on pkg/prowgen/prowgen.go:
```diff
@@ -275,6 +195,7 @@ func generatePeriodicForTest(name string, info *ProwgenInfo, podSpec *corev1.Pod
 		BaseRef: info.Branch,
 	}
 	if pathAlias != nil {
+		base.PathAlias = ""
```
Yeah this was a surprising one, this is actually how periodics have different `JobBase`s than other jobs, when `PathAlias` is present. When I do not clean this up, we generate jobs in o/release differently. I am not exactly sure why it is that way though.

Previously, this was happening because `generatePeriodicForTest` did not pass its `pathAlias` when generating the `JobBase`:
https://github.com/openshift/ci-tools/blob/cee280168ade9363d13d27ad768bc3e5d7c06299/pkg/prowgen/prowgen.go#L267-L268

/hold
I think I can improve this, I agree this is a little hacky and surprising.

petr-muller commented:
@droslean @openshift/test-platform reviews addressed, added a little more code improvements. Changed code in a separate commit so it is reviewable better

petr-muller commented:
/hold cancel

petr-muller commented:
:tada: 

Thanks @droslean
